File: lms-frontend/src/app/services/advanced-search.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface AdvancedSearchRequest {
  query?: string;
  authorIds?: number[];
  categoryIds?: number[];
  yearFrom?: number;
  yearTo?: number;
  availableOnly?: boolean;
  sortBy?: string;
  page?: number;
  size?: number;
}

export interface BookSearchResult {
  content: any[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

@Injectable({
  providedIn: 'root',
})
export class AdvancedSearchService {
  private apiUrl = `${environment.apiBaseUrl}/api/public/search`;

  constructor(private http: HttpClient) {}

  advancedSearch(request: AdvancedSearchRequest): Observable<BookSearchResult> {
    return this.http.post<BookSearchResult>(`${this.apiUrl}/advanced`, request);
  }

  getSimilarBooks(bookId: number, limit: number = 5): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/similar/${bookId}`, {
      params: { limit: limit.toString() },
    });
  }

  getSearchSuggestions(
    query: string,
    limit: number = 10
  ): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/suggestions`, {
      params: { q: query, limit: limit.toString() },
    });
  }

  getAuthorSuggestions(
    query: string,
    limit: number = 10
  ): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/author-suggestions`, {
      params: { q: query, limit: limit.toString() },
    });
  }

  getPopularSearches(limit: number = 10): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/popular`, {
      params: { limit: limit.toString() },
    });
  }
}

File: lms-backend/src/main/java/com/ibizabroker/lms/service/AdvancedSearchService.java
package com.ibizabroker.lms.service;

import com.ibizabroker.lms.dao.BooksRepository;
import com.ibizabroker.lms.entity.Books;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Feature 2: Advanced Search Service
 */
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class AdvancedSearchService {

    private final BooksRepository booksRepository;

    /**
     * Tìm kiếm nâng cao với nhiều tiêu chí
     */
    public Page<Books> advancedSearch(
            String query,
            List<Integer> authorIds,
            List<Integer> categoryIds,
            Integer yearFrom,
            Integer yearTo,
            Boolean availableOnly,
            String sortBy,
            Pageable pageable) {
        
        return booksRepository.advancedSearch(
                query, authorIds, categoryIds, yearFrom, yearTo, 
                availableOnly != null && availableOnly, pageable);
    }

    /**
     * Gợi ý sách tương tự (dựa trên thể loại và tác giả)
     */
    @Cacheable(value = "similar-books", key = "#bookId + '|' + #limit")
    public List<Books> getSimilarBooks(Integer bookId, int limit) {
        Books book = booksRepository.findById(bookId).orElse(null);
        if (book == null) return Collections.emptyList();

        List<Integer> categoryIds = book.getCategories().stream()
                .map(c -> c.getId())
                .collect(Collectors.toList());
        
        if (categoryIds.isEmpty()) return Collections.emptyList();

        return booksRepository.findSimilarBooks(bookId, categoryIds, PageRequest.of(0, limit));
    }

    /**
     * Autocomplete cho tìm kiếm
     */
    @Cacheable(value = "search-suggestions", key = "#query.toLowerCase() + '|' + #limit")
    public List<String> getSearchSuggestions(String query, int limit) {
        if (query == null || query.length() < 2) {
            return Collections.emptyList();
        }
        return booksRepository.findSearchSuggestions(query.toLowerCase(), PageRequest.of(0, limit));
    }

    /**
     * Gợi ý tác giả
     */
    @Cacheable(value = "author-suggestions", key = "#query.toLowerCase() + '|' + #limit")
    public List<String> getAuthorSuggestions(String query, int limit) {
        if (query == null || query.length() < 2) {
            return Collections.emptyList();
        }
        return booksRepository.findAuthorSuggestions(query.toLowerCase(), PageRequest.of(0, limit));
    }

    /**
     * Thống kê tìm kiếm phổ biến (có thể lưu vào Redis)
     */
    private final Map<String, Integer> searchStats = new LinkedHashMap<>();

    public void recordSearch(String query) {
        if (query != null && !query.trim().isEmpty()) {
            searchStats.merge(query.toLowerCase().trim(), 1, (a, b) -> a + b);
        }
    }

    public List<String> getPopularSearches(int limit) {
        return searchStats.entrySet().stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .limit(limit)
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
    }
}
